File: firebase/functions/lib/community/update_membership.dart
import 'package:enum_to_string/enum_to_string.dart';
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/infra/firestore_utils.dart';
import '../utils/subscription_plan_util.dart';
import '../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/utils/utils.dart';

/// Change membership for a user, taking into account the permissions of the requester and
/// potentially the admin count quota of their plan.
class UpdateMembership extends OnCallMethod<UpdateMembershipRequest> {
  UpdateMembership()
      : super(
          'updateMembership',
          (json) => UpdateMembershipRequest.fromJson(json),
        );

  @override
  Future<void> action(
    UpdateMembershipRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);
    orElseInvalidArgument(request.status != null);

    final authUid = context.authUid;
    final communityId = request.communityId;
    final targetUserId = request.userId;
    final targetStatus = request.status;

    // Get needed data.
    final communitySnapshot =
        await firestore.document('community/$communityId').get();
    orElseNotFound(communitySnapshot.exists);
    final community = Community.fromJson(
      firestoreUtils.fromFirestoreJson(communitySnapshot.data.toMap()),
    );

    final requesterMembershipDoc = firestore
        .document('memberships/$authUid/community-membership/$communityId');
    final requesterMembershipSnapshot = await requesterMembershipDoc.get();
    final requesterStatus = requesterMembershipSnapshot.exists
        ? Membership.fromJson(
              firestoreUtils.fromFirestoreJson(
                requesterMembershipSnapshot.data.toMap(),
              ),
            ).status ??
            MembershipStatus.nonmember
        : MembershipStatus.nonmember;

    final targetMembershipDoc = firestore.document(
      'memberships/$targetUserId/community-membership/$communityId',
    );
    final targetMembershipSnapshot = await targetMembershipDoc.get();
    final currentStatus = targetMembershipSnapshot.exists
        ? Membership.fromJson(
              firestoreUtils
                  .fromFirestoreJson(targetMembershipSnapshot.data.toMap()),
            ).status ??
            MembershipStatus.nonmember
        : MembershipStatus.nonmember;

    // Check if user is privileged to attempt this change.
    orElseUnauthorized(
      _isValidMemberUpdate(
        community: community,
        requesterStatus: requesterStatus,
        currentStatus: currentStatus,
        targetStatus: targetStatus!,
        requestingUserId: context.authUid!,
        targetUserId: targetUserId,
      ),
    );

    // If elevating to admin, ensure sufficient quota.

    final adminStatuses = [
      MembershipStatus.admin,
      MembershipStatus.owner,
      MembershipStatus.mod,
    ];
    final facilitatorStatuses = [MembershipStatus.facilitator];
    final elevatingToAdmin = !adminStatuses.contains(currentStatus) &&
        adminStatuses.contains(request.status);
    final elevatingToFacilitator =
        !facilitatorStatuses.contains(currentStatus) &&
            facilitatorStatuses.contains(request.status);

    if (elevatingToAdmin || elevatingToFacilitator) {
      final capabilities =
          await subscriptionPlanUtil.calculateCapabilities(request.communityId);

      if (elevatingToAdmin) {
        final adminMemberships = await firestore
            .collectionGroup('community-membership')
            .where(Membership.kFieldCommunityId, isEqualTo: request.communityId)
            .where(
              Membership.kFieldStatus,
              isEqualTo: EnumToString.convertToString(MembershipStatus.admin),
            )
            .get();
        final ownerMemberships = await firestore
            .collectionGroup('community-membership')
            .where(Membership.kFieldCommunityId, isEqualTo: request.communityId)
            .where(
              Membership.kFieldStatus,
              isEqualTo: EnumToString.convertToString(MembershipStatus.owner),
            )
            .get();
        final modMemberships = await firestore
            .collectionGroup('community-membership')
            .where(Membership.kFieldCommunityId, isEqualTo: request.communityId)
            .where(
              Membership.kFieldStatus,
              isEqualTo: EnumToString.convertToString(MembershipStatus.mod),
            )
            .get();

        final visibleAdmins = [
          ...adminMemberships.documents,
          ...ownerMemberships.documents,
          ...modMemberships.documents,
        ]
            .where(
              (element) =>
                  !(element.data.getBool(Membership.kFieldInvisible) ?? false),
            )
            .toList();

        final currentNumAdmins = visibleAdmins.length;

        if (currentNumAdmins >= (capabilities.adminCount ?? 0)) {
          throw HttpsError(
            HttpsError.resourceExhausted,
            'Insufficient admin count quota.',
            null,
          );
        }
      }

      if (elevatingToFacilitator) {
        final facilitatorMemberships = await firestore
            .collectionGroup('community-membership')
            .where(Membership.kFieldCommunityId, isEqualTo: request.communityId)
            .where(
              Membership.kFieldStatus,
              isEqualTo:
                  EnumToString.convertToString(MembershipStatus.facilitator),
            )
            .get();

        final currentNumFacilitators = facilitatorMemberships.documents
            .where(
              (element) =>
                  !(element.data.getBool(Membership.kFieldInvisible) ?? false),
            )
            .length;

        if (currentNumFacilitators >= (capabilities.facilitatorCount ?? 0)) {
          throw HttpsError(
            HttpsError.resourceExhausted,
            'Insufficient facilitator count quota.',
            null,
          );
        }
      }
    }

    // Make membership change.
    Membership updatedMembership = Membership(
      userId: request.userId,
      communityId: request.communityId,
      status: request.status,
      firstJoined: DateTime.now(),
    );

    if (targetMembershipSnapshot.exists) {
      final data = UpdateData.fromMap(
        jsonSubset(
          [
            Membership.kFieldStatus,
          ],
          firestoreUtils.toFirestoreJson(updatedMembership.toJson()),
        ),
      );
      await targetMembershipDoc.updateData(data);
    } else {
      final data = DocumentData.fromMap(
        firestoreUtils.toFirestoreJson(updatedMembership.toJson()),
      );
      await targetMembershipDoc.setData(data, SetOptions(merge: true));
    }
  }

  /// Determine if requesting user is allowed to do this (ignoring admin quota limitations).
  bool _isValidMemberUpdate({
    required Community community,
    required MembershipStatus requesterStatus,
    required MembershipStatus currentStatus,
    required MembershipStatus targetStatus,
    required String requestingUserId,
    required String targetUserId,
  }) {
    return targetUserId == requestingUserId &&
            _canChangeSelfStatus(
              community: community,
              targetStatus: targetStatus,
              targetUserId: targetUserId,
            ) ||
        _canChangeElevatedMembership(
          requesterStatus: requesterStatus,
          currentStatus: currentStatus,
          targetStatus: targetStatus,
        );
  }

  bool _canChangeElevatedMembership({
    required MembershipStatus requesterStatus,
    required MembershipStatus currentStatus,
    required MembershipStatus targetStatus,
  }) {
    return currentStatus != MembershipStatus.nonmember &&
        (requesterStatus == MembershipStatus.owner ||
            (requesterStatus == MembershipStatus.admin &&
                ![targetStatus, currentStatus]
                    .contains(MembershipStatus.owner)));
  }

  bool _requiresApprovalToJoin(Community community) {
    return community.settingsMigration.requireApprovalToJoin;
  }

  bool _isCommunityCreator(Community community, String userId) {
    return community.creatorId == userId;
  }

  bool _canChangeSelfStatus({
    required Community community,
    required MembershipStatus targetStatus,
    required String targetUserId,
  }) {
    final requiresApproval = _requiresApprovalToJoin(community);
    final selfStatuses = [
      MembershipStatus.attendee,
      MembershipStatus.nonmember,
      if (!requiresApproval) MembershipStatus.member,
    ];
    return selfStatuses.contains(targetStatus) ||
        (targetStatus == MembershipStatus.owner &&
            _isCommunityCreator(community, targetUserId));
  }
}

File: firebase/functions/src/community/update_membership.ts
import * as functions from 'firebase-functions';
import { HttpsError } from 'firebase-functions/lib/providers/https';
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils';
import { OnCallMethod } from '../on_call_function';
import {
  UpdateMembershipRequest, Membership, MembershipStatus, Community,
} from '../types';
import { orElseUnauthorized, orElseInvalidArgument, orElseNotFound, jsonSubset } from '../utils/utils';
import { subscriptionPlanUtil } from '../utils/subscription_plan_util';
import {
  membershipIsAdmin, membershipIsMod, membershipIsMember, membershipIsFacilitator
} from '../types';

export class UpdateMembership extends OnCallMethod<UpdateMembershipRequest> {
  constructor() {
    super('updateMembership', (json) => json as UpdateMembershipRequest);
  }

  async action(
    request: UpdateMembershipRequest,
    context: functions.https.CallableContext
  ): Promise<void> {
    orElseUnauthorized(context.auth?.uid != null);
    orElseInvalidArgument(request.status != null);

    const authUid = context.auth!.uid;
    const { communityId, userId: targetUserId, status: targetStatus } = request;

    const communitySnapshot = await firestore.doc(`community/${communityId}`).get();
    orElseNotFound(communitySnapshot.exists);
    const community = communitySnapshot.data() as Community;

    const requesterMembershipSnap = await firestore
      .doc(`memberships/${authUid}/community-membership/${communityId}`)
      .get();
    const requesterStatus = requesterMembershipSnap.exists
      ? ((requesterMembershipSnap.data() as Membership).status ?? MembershipStatus.nonmember)
      : MembershipStatus.nonmember;

    const targetMembershipDoc = firestore.doc(`memberships/${targetUserId}/community-membership/${communityId}`);
    const targetMembershipSnap = await targetMembershipDoc.get();
    const currentStatus = targetMembershipSnap.exists
      ? ((targetMembershipSnap.data() as Membership).status ?? MembershipStatus.nonmember)
      : MembershipStatus.nonmember;

    orElseUnauthorized(
      this._isValidMemberUpdate({
        community,
        requesterStatus,
        currentStatus,
        targetStatus: targetStatus!,
        requestingUserId: authUid,
        targetUserId,
      })
    );

    const adminStatuses = [MembershipStatus.admin, MembershipStatus.owner, MembershipStatus.moderator];
    const facilitatorStatuses = [MembershipStatus.facilitator];
    const elevatingToAdmin = !adminStatuses.includes(currentStatus) && adminStatuses.includes(targetStatus!);
    const elevatingToFacilitator = !facilitatorStatuses.includes(currentStatus) && facilitatorStatuses.includes(targetStatus!);

    if (elevatingToAdmin || elevatingToFacilitator) {
      const capabilities = await subscriptionPlanUtil.calculateCapabilities(communityId);

      if (elevatingToAdmin) {
        const [adminDocs, ownerDocs, modDocs] = await Promise.all([
          firestore.collectionGroup('community-membership').where('communityId', '==', communityId).where('status', '==', MembershipStatus.admin).get(),
          firestore.collectionGroup('community-membership').where('communityId', '==', communityId).where('status', '==', MembershipStatus.owner).get(),
          firestore.collectionGroup('community-membership').where('communityId', '==', communityId).where('status', '==', MembershipStatus.moderator).get(),
        ]);

        const visibleAdmins = [...adminDocs.docs, ...ownerDocs.docs, ...modDocs.docs]
          .filter((doc) => !(doc.data() as Membership).invisible);

        if (visibleAdmins.length >= (capabilities.adminCount ?? 0)) {
          throw new HttpsError('resource-exhausted', 'Insufficient admin count quota.');
        }
      }

      if (elevatingToFacilitator) {
        const facilitatorDocs = await firestore
          .collectionGroup('community-membership')
          .where('communityId', '==', communityId)
          .where('status', '==', MembershipStatus.facilitator)
          .get();

        const currentNumFacilitators = facilitatorDocs.docs.filter(
          (doc) => !(doc.data() as Membership).invisible
        ).length;

        if (currentNumFacilitators >= (capabilities.facilitatorCount ?? 0)) {
          throw new HttpsError('resource-exhausted', 'Insufficient facilitator count quota.');
        }
      }
    }

    const updatedMembership: Membership = {
      userId: targetUserId,
      communityId,
      status: targetStatus,
      firstJoined: new Date(),
    };

    const data = jsonSubset(
      ['status'],
      firestoreUtils.toFirestoreJson(updatedMembership as unknown as Record<string, unknown>)
    );

    if (targetMembershipSnap.exists) {
      await targetMembershipDoc.update(data);
    } else {
      await targetMembershipDoc.set(
        firestoreUtils.toFirestoreJson(updatedMembership as unknown as Record<string, unknown>),
        { merge: true }
      );
    }
  }

  private _isValidMemberUpdate(opts: {
    community: Community;
    requesterStatus: MembershipStatus;
    currentStatus: MembershipStatus;
    targetStatus: MembershipStatus;
    requestingUserId: string;
    targetUserId: string;
  }): boolean {
    const { requesterStatus, currentStatus, targetStatus, requestingUserId, targetUserId } = opts;

    // Can't modify self (except leaving)
    if (requestingUserId === targetUserId && targetStatus !== MembershipStatus.nonmember) {
      return false;
    }

    // Need to be at least a mod to change others
    if (!membershipIsMod(requesterStatus)) return false;

    // Can't promote to a level equal to or above their own
    if (membershipIsAdmin(targetStatus) && !membershipIsAdmin(requesterStatus)) return false;
    if (targetStatus === MembershipStatus.owner) return false;

    // Can't demote someone at or above your level
    if (membershipIsAdmin(currentStatus) && !membershipIsAdmin(requesterStatus)) return false;

    return true;
  }
}
